From 331dbf4f4bff53c33d6d33ebfead967d300d6554 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Sat, 21 Jun 2025 16:01:27 +0800
Subject: [PATCH 16/24] AOSCOS: gpio: loongson-64bit: Add LS7A GPIO interrupt
 support

On IPASON NL38-N11 laptops, the touchpad device (a HID multitouch device
connected via I2C) uses GPIO 50 to send interrupts. Unfortunately,
gpio-loongson-64bit does not support interrupt for LS7A yet, causing the
touchpad fail to be probed with:

    i2c_hid_acpi i2c-PNP0C50:00: HID over i2c has not been provided an Int IRQ

Add the offset of the GPIO_INT_EN controlling register to
loongson_gpio_ls7a_data to support the interrupt.

And, in LS7A the GPIO 4 ... 56 share the same interrupt pin, we need to
take this fact into account when calling platform_get_irq or we'd fail
with "IRQ index 50 not found."

Signed-off-by: Xi Ruoyao <xry111@xry111.site>

Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>

[Mingcong Bai: Resolved a minor merge conflict since 6.18 in
 drivers/gpio/gpio-loongson-64bit.c.]

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 drivers/gpio/gpio-loongson-64bit.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/gpio/gpio-loongson-64bit.c b/drivers/gpio/gpio-loongson-64bit.c
index 02f181cb2..140c9d8e0 100644
--- a/drivers/gpio/gpio-loongson-64bit.c
+++ b/drivers/gpio/gpio-loongson-64bit.c
@@ -39,6 +39,7 @@ struct loongson_gpio_chip_data {
 	unsigned int		intr_num;
 	irq_flow_handler_t	irq_handler;
 	const struct irq_chip	*girqchip;
+	bool			is_ls7a;
 };
 
 struct loongson_gpio_chip {
@@ -145,6 +146,9 @@ static int loongson_gpio_to_irq(struct gpio_chip *chip, unsigned int offset)
 		writeb(1, lgpio->reg_base + lgpio->chip_data->inten_offset + offset);
 	}
 
+	if (lgpio->chip_data->is_ls7a)
+		offset = MIN(offset, 4);
+
 	return platform_get_irq(pdev, offset);
 }
 
@@ -438,6 +442,7 @@ static const struct loongson_gpio_chip_data loongson_gpio_ls7a_data = {
 	.in_offset = 0xa00,
 	.out_offset = 0x900,
 	.inten_offset = 0xb00,
+	.is_ls7a = true,
 };
 
 /* LS7A2000 chipset GPIO */
-- 
2.51.2

